Library/Homebrew/utils/github.rb

```diff
@@ -438,6 +440,38 @@ def dispatch_event(user, repo, event, **payload)
                   scopes:         CREATE_ISSUE_FORK_OR_PR_SCOPES)
   end
 
+  def fetch_artifact(user, repo, pr, workflow_id, dir)
+    base_url = "#{API_URL}/repos/#{user}/#{repo}"
+    pr_payload = open_api("#{base_url}/pulls/#{pr}")
+    pr_sha = pr_payload["head"]["sha"]
+    pr_branch = pr_payload["head"]["ref"]
+
+    workflow = open_api("#{base_url}/actions/workflows/#{workflow_id}/runs?branch=#{pr_branch}&status=completed")
+    workflow_run = workflow["workflow_runs"].select do |run|
+      run["head_sha"] == pr_sha
+    end
+
+    artifacts = open_api(workflow_run.first["artifacts_url"])
+    onoe "More than one artifact available for PR ##{pr}" if artifacts["total_count"] > 1
+    artifact_url = artifacts["artifacts"].first["archive_download_url"]
+
+    token, username = api_credentials
+    case api_credentials_type
+    when :env_username_password, :keychain_username_password
+      curl_args = { user: "#{username}:#{token}" }
+    when :env_token
+      curl_args = { header: "Authorization: token #{token}" }
+    end
+
+    FileUtils.chdir dir do
+      curl_args[:cache] = Pathname.new(dir)
+      curl_args[:secrets] = [token]
+      downloader = CurlDownloadStrategy.new(artifact_url, "artifact", pr, **curl_args)
+      downloader.fetch
+      downloader.stage
```

Comment:
Can you explain the `stage` here? Not sure I understand why/the format of what's being downloaded.

Reply:
The artefact is a single archive which itself contains the bottle archives and JSONs.

Reply:
Gotcha. In that case I think I'd probably prefer using `curl` and `system "tar"`; this way it's not quite obvious what's going on and why.

Reply:
I've added some comments to clarify what's going on here. I would prefer to use `download_strategy` as this stores the artifact archive in the Homebrew cache, meaning that a second invocation of `brew pr-pull` won't re-download a (possibly huge) set of bottles.

Reply:
Ah, cool. Good enough reason for me 👍

Reply:
Oh, only issue with that: can you confirm it gets handled by `brew cleanup`?

Reply:
Yeah, already tested it locally. It gets scrubbed by `cleanup_unreferenced_downloads`